Assets/Scripts/Level/PuzzleManager.cs

```diff
@@ -0,0 +1,166 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Events;
+using Photon.Pun;
+
+using PlayerType = RoleSelectManager.PlayerType;
+
+public class PuzzleManager : MonoBehaviour
+{
+    [Header("Runes")]
+
+    [SerializeField] public Sprite[] runeSprites;
+    [SerializeField] private SpriteRenderer[] answerRunes;
+    [SerializeField] private RuneInteractable[] runeRocks;
+
+    [Header("Events")]
+
+    [SerializeField] private UnityEvent puzzleCompleteEvent;
+
+    [Header("Photon")]
+
+    [SerializeField] private PhotonView photonView;
+
+    private int answerLength;
+    private int[] runeAnswer;
+
+    private void OnValidate()
+    {
+        if (answerRunes.Length > runeSprites.Length)
+        {
+            Debug.LogWarning("There are not enough unique sprites to support the answer length!");
+        }
+    }
+
+    private void Start()
+    {
+        answerLength = answerRunes.Length;
+
+        if (PhotonNetwork.IsMasterClient)
+        {
+            GenerateRandomAnswer();
+            photonView.RPC("RPC_SetPuzzleAnswer", RpcTarget.All, runeAnswer);
+        }
+
+        HidePuzzleHalf();
+    }
+
+    private void OnEnable()
+    {
+        foreach (RuneInteractable runeRock in runeRocks)
+        {
+            runeRock.RuneUpdateEvent.AddListener(HandleRuneUpdate);
+        }
+    }
+
+    private void OnDisable()
+    {
+        foreach (RuneInteractable runeRock in runeRocks)
+        {
+            runeRock.RuneUpdateEvent.RemoveListener(HandleRuneUpdate);
+        }
+    }
+
+    private void GenerateRandomAnswer()
+    {
+        runeAnswer = new int[answerLength];
+        List<int> values = new List<int>();
+
+        for (int i = 0; i < runeSprites.Length; i++)
+        {
+            values.Add(i);
+        }
+
+        for (int i = 0; i < answerLength; i++)
+        {
+            int randomIndex = Random.Range(0, values.Count);
+            runeAnswer[i] = values[randomIndex];
+
+            values.RemoveAt(randomIndex);
```

Comment:
A `HashSet` could be used to store the `values` while generating, so that removing a value after randomly selecting it costs `O(1)` instead of `O(n)` in the worst case.

Reply:
I don't think this is feasible because I cannot find a way to randomly pick an element from the hashset.

Reply:
That’s true, it’d probably have to be a Dictionary. But this might be overly complex for a simple operation, especially since there probably won’t ever be a huge number of runes making up the answer.